### PR TITLE
distsql: basic planning for merge joins

### PR DESCRIPTION
**Pull request by RaduBerinde**
We detect when the two sides of the join have matching orderings on all equality
columns and use the merge joiner.

Note that as things stand, the join node does not guarantee any output ordering.
This means that we may have an unnecessary sorter after a merge join, or - much
more importantly - the output of a merge join won't be used as the input of
another merge join.

Updates #16580.

I have not looked at the running time of a merge join vs hash join on a large dataset (TPC-H). I was hoping @arjunravinarayan could help with that. Even a single-node deployment would give us a good idea of the speed difference.

### CODE REVIEW

**Comment by arjunravinarayan** [comment]
:lgtm:

Hold off on merge until I experimentally deploy and confirm the speedup on TPC-H?

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 1395 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-KpzX5rv6StGCYBQCosv:-KpzX5rwEaA58EzOVXz7:ba6o9yf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c157f1d32721b69dc873000dcc56703a0a80e806/pkg/sql/distsql_physical_planner.go#L1395)):*
> ```Go
> 			// TODO(radu): we currently only use merge joins when we have an ordering on
> 			// all equality columns. We should relax this by either:
> 			//  - implementing a hybrid hash/merge processor which implements merge
> ```

Well currently the `HashJoiner` as implemented will take advantage of incomplete orderings, by doing the hybrid thing you suggest. Am I missing something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-:-Kq-qz5qhwJpp5fpNK2p:b48usnt)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/distsql_physical_planner.go, line 1395 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-KpzX5rv6StGCYBQCosv:-Kq-sIUMBVTNm08D5Dnn:b-puar1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c157f1d32721b69dc873000dcc56703a0a80e806/pkg/sql/distsql_physical_planner.go#L1395)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Well currently the `HashJoiner` as implemented will take advantage of incomplete orderings, by doing the hybrid thing you suggest. Am I missing something?
</blockquote></details>

It does none of that..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214)*

**Comment by arjunravinarayan** [comment]
*[pkg/sql/distsql_physical_planner.go, line 1395 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-KpzX5rv6StGCYBQCosv:-Kq-ti_Q97aKp5x7WdUp:b6tiu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c157f1d32721b69dc873000dcc56703a0a80e806/pkg/sql/distsql_physical_planner.go#L1395)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It does none of that..
</blockquote></details>

You are right, of course. I checked the code, and it's obviously not (and I've looked at that code recently). I am getting confused by my interview question, as I've now merged in-my-head all the various optimizations I've gone through in my interviews. Sorry.

We should absolutely implement a hybrid hash/merge processor at some point, this is a good todo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214)*

**Comment by arjunravinarayan** [comment]
Reviewed 6 of 6 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214)*

**Comment by arjunravinarayan** [comment]
*[pkg/sql/distsql_physical_planner.go, line 1395 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-KpzX5rv6StGCYBQCosv:-Kq0CDnLyIJUa8NRsKJw:b9i0vld) ([raw file](https://github.com/cockroachdb/cockroach/blob/c157f1d32721b69dc873000dcc56703a0a80e806/pkg/sql/distsql_physical_planner.go#L1395)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

You are right, of course. I checked the code, and it's obviously not (and I've looked at that code recently). I am getting confused by my interview question, as I've now merged in-my-head all the various optimizations I've gone through in my interviews. Sorry.

We should absolutely implement a hybrid hash/merge processor at some point, this is a good todo.
</blockquote></details>

I realized on my bike ride home why I was confused. We do this in the `exceptAll` processor in `algebraic_set_op.go`. We have all the pieces in place to do this for the hashjoiner as well: the common functionality is in `streamGroupAccumulator`, and we can use it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214)*

**Comment by RaduBerinde** [comment]
Yes, I plan to wait for that before merging.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 1395 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-KpzX5rv6StGCYBQCosv:-Kq12IVAP48zlXNWfOnK:b-gtv5gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c157f1d32721b69dc873000dcc56703a0a80e806/pkg/sql/distsql_physical_planner.go#L1395)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I realized on my bike ride home why I was confused. We do this in the `exceptAll` processor in `algebraic_set_op.go`. We have all the pieces in place to do this for the hashjoiner as well: the common functionality is in `streamGroupAccumulator`, and we can use it.
</blockquote></details>

Ah, indeed, very good point!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-:-Kq12DtufhEqZFBhbhS2:b-avmz6g)*

**Comment by knz** [comment]
cc @cockroachdb/sql-planning 

Sorry that I am late to the discussion. Out of curiosity, was there a reason to place the computation of the merge join ordering (and the decision to use a merge join) in the physical planning stage, as opposed to earlier? I would certainly see this information useful to be annotated in the planNode earlier than that, ready for inspection by EXPLAIN.

**Comment by RaduBerinde** [comment]
There is no planNode that can do a merge join.. Are you suggesting augmenting the current joinNode with ordering information (basically pretend it can do a merge join even if it can't)?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-:-Kq34Y47EpTHqc7ZfEUe:b-ssu5ws)*

**Comment by knz** [comment]
>  Are you suggesting
> augmenting the current joinNode with ordering information (basically
> pretend it can do a merge join even if it can't)?
> 

Yes that's what I mean.


-- 
Raphael 'kena' Poss

**Comment by RaduBerinde** [comment]
I see, I will look at doing that.

**Comment by arjunravinarayan** [comment]
TL;DR: 20s merge join VS TODO hash join. Ship it.

I ran this on a 6 node cluster, on a fairly artificial query (one that in fact revealed another issue, filed at #17270). I toggled the `merge_joins.enabled` cluster setting only:

The first run is with merge joins enabled.
```
cockroach@cockroach-cambridge-0007:~$ time ./cockroach sql --url 'postgresql://root@cockroach-cambridge-0004:26257?sslmode=disable' -e 'SELECT count(l_orderkey) FROM tpch.orders@o_ok, tpch.lineitem@l_ok WHERE orders.o_orderkey = lineitem.l_orderkey'
# Client version: CockroachDB CCL 331d89d (linux amd64, built 2017/07/24 18:49:48, go1.8.3)
# Server version: CockroachDB CCL c157f1d-dirty (linux amd64, built 2017/07/27 14:38:10, go1.8.3)
# Cluster ID: b76701ed-8c3c-49f0-90ab-0ac45c446bb1
+-------------------+
| count(l_orderkey) |
+-------------------+
|           6001215 |
+-------------------+
(1 row)

real	0m19.055s
user	0m0.032s
sys	0m0.020s
```

The second run is with merge joins disabled has not finished, and it's been over 20 minutes. I now think there is a bug in the hashjoiner implementation. Nevertheless, the mergejoiner works well on large queries, and we should merge this code.

This query arguably does not test the performance discrepancy at the maximum: the index join in advance of the merge join is still potentially the gating factor.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-:-Kq51SUcwGMGvVRMu4qZ:b-x8ln32)*

**Comment by RaduBerinde** [comment]
Thanks Arjun! I will try the refactoring that Raphael suggested, but it won't affect the execution itself.

**Comment by RaduBerinde** [comment]
@knz Moved the merge join ordering in the joinNode and added it to `EXPLAIN` (plus some testcases with that). Much nicer, thanks for the suggestion!

**Comment by knz** [comment]
:lgtm:

---

Reviewed 2 of 6 files at r1, 7 of 7 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214#-:-KqNC_R0WhExukucSFjN:bnfp4nl)*

**Comment by arjunravinarayan** [comment]
Reviewed 7 of 7 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17214)*
